# MLOps Training 2026/2027 — Task 2
## Notebook 1: Read and Join the Tables

In this notebook, I will read the Olist dataset tables from the PostgreSQL database and check their structure before joining them.

The main goal is to create one table with one row per order that can be used in the next notebooks.

I will check the number of rows, keys, duplicates, and what each table represents. Tables such as order items and order payments contain multiple rows for the same order, so they need to be aggregated before joining.

In [64]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(
    "postgresql+psycopg2://postgres:afnan@localhost:5432/olist"
)

print("Connected successfully!")

Connected successfully!


## 1. Read the Database Tables
First, I will check which tables are available in the PostgreSQL database and then load them into pandas DataFrames.

In [65]:
query = """
SELECT table_name
FROM information_schema.tables
WHERE table_schema = 'public'
ORDER BY table_name;
"""

tables = pd.read_sql(query, engine)
tables

,table_name
0,customers
1,geolocation
2,order_items
3,order_payments
4,order_reviews
5,orders
6,product_category_name_translation
7,products
8,sellers


### Load the Tables
Now I will load the nine database tables into pandas DataFrames so I can inspect and work with them.

In [66]:
customers = pd.read_sql("SELECT * FROM customers;", engine)
geolocation = pd.read_sql("SELECT * FROM geolocation;", engine)
order_items = pd.read_sql("SELECT * FROM order_items;", engine)
order_payments = pd.read_sql("SELECT * FROM order_payments;", engine)
order_reviews = pd.read_sql("SELECT * FROM order_reviews;", engine)
orders = pd.read_sql("SELECT * FROM orders;", engine)
products = pd.read_sql("SELECT * FROM products;", engine)
sellers = pd.read_sql("SELECT * FROM sellers;", engine)
product_category = pd.read_sql(
    "SELECT * FROM product_category_name_translation;",
    engine
)

print("All tables loaded successfully!")

All tables loaded successfully!


## 2. Inspect the Tables
Before joining the tables, I will check their size and structure. This helps identify the main keys and tables that may contain more than one row for the same order.

In [67]:
dataframes = {
    "customers": customers,
    "geolocation": geolocation,
    "order_items": order_items,
    "order_payments": order_payments,
    "order_reviews": order_reviews,
    "orders": orders,
    "products": products,
    "sellers": sellers,
    "product_category": product_category
}

table_summary = pd.DataFrame({
    "table": dataframes.keys(),
    "rows": [df.shape[0] for df in dataframes.values()],
    "columns": [df.shape[1] for df in dataframes.values()]
})

table_summary

,table,rows,columns
0,customers,99441,5
1,geolocation,1000163,5
2,order_items,112650,7
3,order_payments,103886,5
4,order_reviews,99224,7
5,orders,99441,8
6,products,32951,9
7,sellers,3095,4
8,product_category,71,2


### 2.1 Keys and Duplicates
The tables have different numbers of rows, especially `order_items` and `order_payments`. I will check the main keys and whether the same order appears more than once before joining the tables.

In [68]:
key_checks = pd.DataFrame({
    "table": ["orders", "customers", "order_items", "order_payments", "order_reviews"],
    "key": ["order_id", "customer_id", "order_id", "order_id", "order_id"],
    "rows": [
        len(orders),
        len(customers),
        len(order_items),
        len(order_payments),
        len(order_reviews)
    ],
    "unique_keys": [
        orders["order_id"].nunique(),
        customers["customer_id"].nunique(),
        order_items["order_id"].nunique(),
        order_payments["order_id"].nunique(),
        order_reviews["order_id"].nunique()
    ],
    "duplicated_keys": [
        orders["order_id"].duplicated().sum(),
        customers["customer_id"].duplicated().sum(),
        order_items["order_id"].duplicated().sum(),
        order_payments["order_id"].duplicated().sum(),
        order_reviews["order_id"].duplicated().sum()
    ]
})

key_checks

,table,key,rows,unique_keys,duplicated_keys
0,orders,order_id,99441,99441,0
1,customers,customer_id,99441,99441,0
2,order_items,order_id,112650,98666,13984
3,order_payments,order_id,103886,99440,4446
4,order_reviews,order_id,99224,98673,551


The `orders` table has one unique `order_id` for every row, so it can be used as the base table.

`order_items` and `order_payments` contain repeated `order_id` values because one order can have multiple items or payment records. These tables need to be aggregated before joining them with `orders`.

The `order_reviews` table also contains some repeated order IDs, so I will handle it carefully if it is included in the final join.

In [69]:
for name, df in dataframes.items():
    print(f"\n{name.upper()}")
    display(df.head(2))


CUSTOMERS


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,09790,sao bernardo do campo,SP



GEOLOCATION


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP



ORDER_ITEMS


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93



ORDER_PAYMENTS


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39



ORDER_REVIEWS


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10,2018-03-11 03:05:13



ORDERS


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13



PRODUCTS


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0



SELLERS


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP



PRODUCT_CATEGORY


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories


### 2.2 Table Structure
From the sample rows, the tables contain different parts of the order process:

- `orders` contains the main order information and connects each order to a customer.
- `customers` contains customer location information.
- `order_items` contains the products and sellers associated with each order.
- `order_payments` contains payment information for each order.
- `order_reviews` contains customer review information.
- `products` contains product details and categories.
- `sellers` contains seller location information.
- `geolocation` contains location data based on ZIP-code prefixes.
- `product_category` translates product category names from Portuguese to English.

The main table for the ML dataset will be `orders`, since it already has one row per order. The other tables will be joined to it where needed.

In [70]:
multiple_items = (
    order_items["order_id"]
    .value_counts()
    .loc[lambda x: x > 1]
)

print("Orders with multiple item rows:", len(multiple_items))
print("Maximum item rows for one order:", multiple_items.max())

multiple_items.head()

Orders with multiple item rows: 9803
Maximum item rows for one order: 21


order_id
8272b63d03f5f79c56e9e4120aec44ef    21
1b15974a0141d54e36626dca3fdc731a    20
ab14fdcfbe524636d65ee38360e22ce8    20
428a2f660dc84138d969ccd69a0ab6d5    15
9ef13efd6949e4573a18964dd1bbe7f5    15
Name: count, dtype: int64

In [71]:
multiple_payments = (
    order_payments["order_id"]
    .value_counts()
    .loc[lambda x: x > 1]
)

print("Orders with multiple payment rows:", len(multiple_payments))
print("Maximum payment rows for one order:", multiple_payments.max())

multiple_payments.head()

Orders with multiple payment rows: 2961
Maximum payment rows for one order: 29


order_id
fa65dad1b0e818e3ccc5cb0e39231352    29
ccf804e764ed5650cd8759557269dc13    26
285c2e15bebd4ac83635ccc563dc71f4    22
895ab968e7bb0d5659d16cd74cd1650c    21
ee9ca989fc93ba09a6eddc250ce01742    19
Name: count, dtype: int64

### 2.3 Seller Geography
The `order_items` table contains `seller_id`, while the `sellers` table contains seller location information. I will combine them before creating an order-level seller summary.

In [72]:
order_items_sellers = order_items.merge(
    sellers,
    on="seller_id",
    how="left"
)

order_items_sellers[
    [
        "order_id",
        "seller_id",
        "seller_zip_code_prefix",
        "seller_city",
        "seller_state"
    ]
].head()

,order_id,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,00010242fe8c5a6d1ba2dd792cb16214,48436dade18ac8b2bce089ec2a041202,27277,volta redonda,SP
1,00018f77f2f0320c557190d7a144bdd3,dd7ddc04e1b6c2c614352b383efe2d36,3471,sao paulo,SP
2,000229ec398224ef6ca0657da4fc703e,5b51032eddd242adc84c38acab88f23d,37564,borda da mata,MG
3,00024acbcdf0a6daa1e931b038114c75,9d7a1d34a5052409006425275ba1c2b4,14403,franca,SP
4,00042b26cf59d7ce69dfabb4e55b4fd9,df560393f3a51e74553ab94004ba5c87,87900,loanda,PR


### 2.4 Primary Seller Location
Some orders contain more than one seller. To keep one row per order, I will use the seller associated with the first order item as the primary seller location.

The existing `unique_sellers` feature will still show whether an order contains products from multiple sellers.

In [73]:
primary_seller = (
    order_items_sellers
    .sort_values(["order_id", "order_item_id"])
    .groupby("order_id")
    .first()
    .reset_index()
    [
        [
            "order_id",
            "seller_zip_code_prefix",
            "seller_city",
            "seller_state"
        ]
    ]
)

primary_seller.head()

,order_id,seller_zip_code_prefix,seller_city,seller_state
0,00010242fe8c5a6d1ba2dd792cb16214,27277,volta redonda,SP
1,00018f77f2f0320c557190d7a144bdd3,3471,sao paulo,SP
2,000229ec398224ef6ca0657da4fc703e,37564,borda da mata,MG
3,00024acbcdf0a6daa1e931b038114c75,14403,franca,SP
4,00042b26cf59d7ce69dfabb4e55b4fd9,87900,loanda,PR


In [74]:
print("Rows:", len(primary_seller))
print("Unique order IDs:", primary_seller["order_id"].nunique())
print("Duplicate order IDs:", primary_seller["order_id"].duplicated().sum())

Rows: 98666
Unique order IDs: 98666
Duplicate order IDs: 0


### 2.5 Geolocation Coordinates
The `geolocation` table contains latitude and longitude for ZIP-code prefixes. Since a ZIP-code prefix can appear more than once, I will first calculate one average coordinate for each prefix.

These coordinates will later be used to estimate the geographic distance between the customer and the primary seller.

In [75]:
geolocation_agg = (
    geolocation
    .groupby("geolocation_zip_code_prefix")
    .agg(
        latitude=("geolocation_lat", "mean"),
        longitude=("geolocation_lng", "mean")
    )
    .reset_index()
)

print("Rows:", len(geolocation_agg))
print(
    "Duplicate ZIP prefixes:",
    geolocation_agg["geolocation_zip_code_prefix"].duplicated().sum()
)

geolocation_agg.head()

Rows: 19015
Duplicate ZIP prefixes: 0


,geolocation_zip_code_prefix,latitude,longitude
0,1001,-23.550190,-46.634024
1,1002,-23.548146,-46.634979
2,1003,-23.548994,-46.635731
3,1004,-23.549799,-46.634757
4,1005,-23.549456,-46.636733


### 2.6 Customer and Seller Coordinates
I will match the customer and primary seller ZIP-code prefixes with the aggregated geolocation table to add latitude and longitude for both sides of the order.

In [76]:
customer_geo = geolocation_agg.rename(
    columns={
        "geolocation_zip_code_prefix": "customer_zip_code_prefix",
        "latitude": "customer_latitude",
        "longitude": "customer_longitude"
    }
)

seller_geo = geolocation_agg.rename(
    columns={
        "geolocation_zip_code_prefix": "seller_zip_code_prefix",
        "latitude": "seller_latitude",
        "longitude": "seller_longitude"
    }
)

In [77]:
customers["customer_zip_code_prefix"] = (
    customers["customer_zip_code_prefix"].astype(int)
)

primary_seller["seller_zip_code_prefix"] = (
    primary_seller["seller_zip_code_prefix"].astype(int)
)

In [78]:
customers_with_geo = customers.merge(
    customer_geo,
    on="customer_zip_code_prefix",
    how="left"
)

primary_seller_with_geo = primary_seller.merge(
    seller_geo,
    on="seller_zip_code_prefix",
    how="left"
)

print("Customers with missing coordinates:",
      customers_with_geo["customer_latitude"].isna().sum())

print("Primary sellers with missing coordinates:",
      primary_seller_with_geo["seller_latitude"].isna().sum())

Customers with missing coordinates: 278
Primary sellers with missing coordinates: 218


## 3. Aggregate Tables Before Joining
Both `order_items` and `order_payments` can contain multiple rows for the same order. Joining them directly with the `orders` table would create duplicate order rows.

To keep one row per order, I will first aggregate these tables by `order_id`.

In [79]:
order_items_agg = (
    order_items
    .groupby("order_id")
    .agg(
        item_count=("order_item_id", "count"),
        total_price=("price", "sum"),
        total_freight_value=("freight_value", "sum"),
        unique_products=("product_id", "nunique"),
        unique_sellers=("seller_id", "nunique")
    )
    .reset_index()
)

order_items_agg.head()

,order_id,item_count,total_price,total_freight_value,unique_products,unique_sellers
0,00010242fe8c5a6d1ba2dd792cb16214,1,58.90,13.29,1,1
1,00018f77f2f0320c557190d7a144bdd3,1,239.90,19.93,1,1
2,000229ec398224ef6ca0657da4fc703e,1,199.00,17.87,1,1
3,00024acbcdf0a6daa1e931b038114c75,1,12.99,12.79,1,1
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,199.90,18.14,1,1


In [80]:
print("Rows:", len(order_items_agg))
print("Unique order IDs:", order_items_agg["order_id"].nunique())
print("Duplicate order IDs:", order_items_agg["order_id"].duplicated().sum())

Rows: 98666
Unique order IDs: 98666
Duplicate order IDs: 0


In [81]:
order_payments_agg = (
    order_payments
    .groupby("order_id")
    .agg(
        payment_count=("payment_sequential", "count"),
        total_payment_value=("payment_value", "sum"),
        max_installments=("payment_installments", "max"),
        payment_type_count=("payment_type", "nunique")
    )
    .reset_index()
)

order_payments_agg.head()

,order_id,payment_count,total_payment_value,max_installments,payment_type_count
0,00010242fe8c5a6d1ba2dd792cb16214,1,72.19,2,1
1,00018f77f2f0320c557190d7a144bdd3,1,259.83,3,1
2,000229ec398224ef6ca0657da4fc703e,1,216.87,5,1
3,00024acbcdf0a6daa1e931b038114c75,1,25.78,2,1
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,218.04,3,1


In [82]:
print("Rows:", len(order_payments_agg))
print("Unique order IDs:", order_payments_agg["order_id"].nunique())
print("Duplicate order IDs:", order_payments_agg["order_id"].duplicated().sum())

Rows: 99440
Unique order IDs: 99440
Duplicate order IDs: 0


### 3.1 Aggregation Check
After aggregation, both `order_items` and `order_payments` have one row per `order_id`. This means they can now be joined with the `orders` table without creating extra order rows.

## 4. Build the ML Table
I will use the `orders` table as the base because it already contains one row per order.

The customer information and the aggregated order item and payment information will be added to the base table. The number of rows will be checked after the joins to make sure each order still appears only once.

In [83]:
ml_table = (
    orders
    .merge(
        customers_with_geo,
        on="customer_id",
        how="left"
    )
    .merge(
        order_items_agg,
        on="order_id",
        how="left"
    )
    .merge(
        order_payments_agg,
        on="order_id",
        how="left"
    )
    .merge(
        primary_seller_with_geo,
        on="order_id",
        how="left"
    )
)

ml_table.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,unique_sellers,payment_count,total_payment_value,max_installments,payment_type_count,seller_zip_code_prefix,seller_city,seller_state,seller_latitude,seller_longitude
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,1.0,3.0,38.71,1.0,2.0,9350.0,maua,SP,-23.680729,-46.444238
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,...,1.0,1.0,141.46,1.0,1.0,31570.0,belo horizonte,SP,-19.807681,-43.980427
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,1.0,1.0,179.12,3.0,1.0,14840.0,guariba,SP,-21.363502,-48.229601
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,7c142cf63193a1473d2e66489a9ae977,59296,...,1.0,1.0,72.20,1.0,1.0,31842.0,belo horizonte,MG,-19.837682,-43.924053
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,72632f0f9dd73dfee390c9b22eb56dd6,9195,...,1.0,1.0,28.62,1.0,1.0,8752.0,mogi das cruzes,SP,-23.543395,-46.262086


In [84]:
print("Rows in original orders table:", len(orders))
print("Rows in ML table:", len(ml_table))
print("Unique order IDs:", ml_table["order_id"].nunique())
print("Duplicate order IDs:", ml_table["order_id"].duplicated().sum())

Rows in original orders table: 99441
Rows in ML table: 99441
Unique order IDs: 99441
Duplicate order IDs: 0


In [85]:
print("Orders without item information:", ml_table["item_count"].isna().sum())
print("Orders without payment information:", ml_table["payment_count"].isna().sum())
print("Orders without customer information:", ml_table["customer_unique_id"].isna().sum())

Orders without item information: 775
Orders without payment information: 1
Orders without customer information: 0


### 4.1 Join Validation
The joined table still contains 99,441 rows and 99,441 unique order IDs, with no duplicated orders.

All orders matched with customer information. Some orders do not have matching item or payment information, so I will check their order status before deciding how they should be handled.

In [86]:
missing_items_status = (
    ml_table[ml_table["item_count"].isna()]
    ["order_status"]
    .value_counts()
)

missing_items_status

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [87]:
ml_table[ml_table["payment_count"].isna()][
    ["order_id", "order_status"]
]

,order_id,order_status
30710,bfbd0f9bdef84302105ad712db648a6c,delivered


The missing item information is mostly associated with unavailable and canceled orders. There is also one delivered order without matching payment information.

I will keep these rows in the ML table for now. Missing values and any filtering decisions will be handled in the later notebooks.

### 4.2 Customer–Seller Distance
Using the customer and seller latitude and longitude values, I will calculate the approximate straight-line distance between them using the Haversine formula.

In [88]:
import numpy as np

def haversine_distance(lat1, lon1, lat2, lon2):
    earth_radius_km = 6371

    lat1 = np.radians(lat1)
    lon1 = np.radians(lon1)
    lat2 = np.radians(lat2)
    lon2 = np.radians(lon2)

    dlat = lat2 - lat1
    dlon = lon2 - lon1

    a = (
        np.sin(dlat / 2) ** 2
        + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2) ** 2
    )

    c = 2 * np.arcsin(np.sqrt(a))

    return earth_radius_km * c

In [89]:
ml_table["distance_km"] = haversine_distance(
    ml_table["customer_latitude"],
    ml_table["customer_longitude"],
    ml_table["seller_latitude"],
    ml_table["seller_longitude"]
)

print(
    "Orders with missing distance:",
    ml_table["distance_km"].isna().sum()
)

ml_table[
    [
        "order_id",
        "customer_state",
        "seller_state",
        "distance_km"
    ]
].head()

Orders with missing distance: 1266


,order_id,customer_state,seller_state,distance_km
0,e481f51cbdc54678b7cc49136f2d6af7,SP,SP,18.576110
1,53cdb2fc8bc7dce0b6741e2150273451,BA,SP,851.495069
2,47770eb9100c2d0c44946d9cf07ec65d,GO,SP,514.410666
3,949d5b44dbf5de918fe9c16f97b45f8a,RN,MG,1822.226336
4,ad21c59c0840e6cb83a9ceb5573f8159,SP,SP,29.676625


The distance could be calculated for most orders. A small number of orders have missing distance values because customer or seller coordinates were not available.

These missing values will be handled later during feature engineering.

## 5. Save the ML Table
The joined table will be saved as an artifact so it can be loaded directly in Notebook 2 without rebuilding the joins.

In [90]:
ml_table.to_csv(
    "artifacts/ml_table.csv",
    index=False
)

print("ML table saved successfully!")

ML table saved successfully!


## 6. Final Check
Before finishing this notebook, I will confirm that the saved ML table has one row per order.

In [91]:
saved_ml_table = pd.read_csv("artifacts/ml_table.csv")

print("Saved rows:", len(saved_ml_table))
print("Unique order IDs:", saved_ml_table["order_id"].nunique())
print("Duplicate order IDs:", saved_ml_table["order_id"].duplicated().sum())

Saved rows: 99441
Unique order IDs: 99441
Duplicate order IDs: 0


## 7. Conclusion
The Olist tables were loaded and checked before joining. Since `order_items` and `order_payments` can contain multiple rows per order, they were aggregated first.

The final ML table contains one row per order with no duplicate `order_id` values. It has been saved as `ml_table.csv` and will be used as the input for Notebook 2.